SSA/Projects/FullyHomomorphicEncryption/Basic.lean

```diff
@@ -81,72 +142,240 @@ abbrev R := (ZMod q)[X] ⧸ (Ideal.span {f q n})
 -/
 @[simp]
 theorem R.fromPoly_representative : forall a : R q n, (R.fromPoly (n:=n) (R.representative q n a)) = a := by
  intro a
  simp [R.representative]
  rw [Polynomial.modByMonic_eq_sub_mul_div _ (f_monic q n)]
  rw [RingHom.map_sub (R.fromPoly (q := q) (n:=n)) _ _]
- have hker : forall x, fromPoly (f q n * x) = 0 := by
-   intro x
-   unfold fromPoly
-   apply Ideal.Quotient.eq_zero_iff_mem.2
-   rw [Ideal.mem_span_singleton]
-   simp [Dvd.dvd]
-   use x
- rw [hker _]
+ rw [R.fromPoly_kernel_eq_zero]
  simp
  apply Function.surjInv_eq
 
 
 /--
-Characterization theorem for any potential representative.
+Characterization theorem for any potential representative (in terms of ideals).
 For an  `a : (ZMod q)[X]`, the representative of its equivalence class
 is just `a + i` for some `i ∈ (Ideal.span {f q n})`.
 -/
-theorem R.fromPoly_rep'_eq : forall a : (ZMod q)[X], ∃ i ∈ Ideal.span {f q n}, (R.fromPoly (n:=n) a).representative' = a + i := by
+theorem R.fromPoly_rep'_eq_ideal : forall a : (ZMod q)[X], ∃ i ∈ Ideal.span {f q n}, (R.fromPoly (n:=n) a).representative' = a + i := by
   intro a
   exists (R.fromPoly (n:=n) a).representative' - a
   constructor
   · apply Ideal.Quotient.eq.1
     simp [R.representative', Function.surjInv_eq]
   · ring
   done
+/--
+Characterization theorem for any potential representative (in terms of elements).
+For an  `a : (ZMod q)[X]`, the representative of its equivalence class
+is a concrete element of the form `a + k * (f q n)` for some `k ∈ (ZMod q)[X]`.
+-/
+theorem R.fromPoly_rep'_eq_element (a : (ZMod q)[X]) : ∃ (k : (ZMod q)[X]), (R.fromPoly (n:=n) a).representative' = k * (f q n) + a := by
+  have H : ∃ i ∈ Ideal.span {f q n}, (R.fromPoly (n:=n) a).representative' = a + i := by
+    apply R.fromPoly_rep'_eq_ideal
+  obtain ⟨i, iInIdeal, ih⟩ := H
+  have fqn_div_i : (f q n) ∣ i  := by
+    rw[← Ideal.mem_span_singleton]
+    assumption
+  have i_multiple_fqn : ∃ (k : (ZMod q)[X]), i = k * (f q n) := by
+    apply dvd_iff_exists_eq_mul_left.mp
+    assumption
+  obtain ⟨k, hk⟩ := i_multiple_fqn
+  exists k
+  subst hk
+  rw[ih]
+  ring_nf
+
+/-- A theorem similar to `R.fromPoly_rep'_eq_element` but uses `fromPoly.toFun` to be more deterministic,
+  as the `toFun` sometimes sneaks in due to coercions. -/
+theorem R.representatitive'_toFun_fromPoly_eq_element (a : (ZMod q)[X]) : ∃ (k : (ZMod q)[X]),
+  R.representative' q n ((R.fromPoly (q := q) (n := n)).toFun a) = a + k * (f q n) := by
+  have H : ∃ (k : (ZMod q)[X]), (R.fromPoly (n:=n) a).representative' = k * (f q n) + a := by apply
+    R.fromPoly_rep'_eq_element;
+  obtain ⟨k, hk⟩ := H
+  exists k
+  ring_nf at hk ⊢
+  rw[← hk]
+  norm_cast
+
+/-- The representative of 0 wil live in the ideal of {f q n}. To show that such an element is a multiple of {f q n}, use `Ideal.mem_span_singleton'`-/
+theorem R.representative'_zero_ideal : R.representative' q n 0 ∈ Ideal.span {f q n} := by
```

Comment:
Use `representative` and then it is just `0`

Reply:
Yes, that's `R.representative_zero` below, this is just a helper lemma to prove that.

Reply:
Get rid of `representative'`. There's no need for it. You might want to prove the lemma that two polys are equal in the quotient by `r` if `p %m r = q %m r`, which is itself equivalent to `r | p - q`